File: cannabis_management/cannabis_management/page/ar_summary/ar_summary.js
frappe.pages['ar-summary'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Accounts Receivable',
		single_column: true
	});
	wrapper.page = page;

	var METHOD = 'cannabis_management.cannabis_management.page.ar_summary.ar_summary';

	// company / dates are server-side filters (reload); search / status / sort
	// are applied instantly on the loaded rows. Ledger data is fetched lazily.
	var state = {
		view: 'summary',          // 'summary' | 'ledgers'
		company: '', from_date: '', to_date: '',
		search: '', status: 'all', sort: 'outstanding',
		rows: [], ledgers: null, prepared: ''
	};

	page.main.html(`
		<div class="ars-wrap">
			<header class="ars-header">
				<div class="ars-header-text">
					<h1 class="ars-h1">Accounts Receivable</h1>
					<p class="ars-sub" id="ars-prepared"></p>
				</div>
				<div class="ars-header-actions">
					<div class="ars-seg" id="ars-seg">
						<button class="ars-seg-btn is-active" data-view="summary">Summary</button>
						<button class="ars-seg-btn" data-view="ledgers">Ledgers</button>
					</div>
					<button id="ars-refresh" class="ars-icon-btn" title="Refresh"><span id="ars-refresh-ic">&#8635;</span></button>
					<button id="ars-export" class="ars-primary-btn">&#8681;&nbsp;Export to Excel</button>
				</div>
			</header>

			<section class="ars-kpis">
				<div class="ars-kpi ars-kpi--invoice">
					<div class="ars-kpi-top"><span class="ars-kpi-label">Invoice Total</span><span class="ars-kpi-dot"></span></div>
					<div class="ars-kpi-value" id="ars-k-inv">$0.00</div>
				</div>
				<div class="ars-kpi ars-kpi--paid">
					<div class="ars-kpi-top"><span class="ars-kpi-label">Paid Amount</span><span class="ars-kpi-dot"></span></div>
					<div class="ars-kpi-value" id="ars-k-paid">$0.00</div>
				</div>
				<div class="ars-kpi ars-kpi--out">
					<div class="ars-kpi-top"><span class="ars-kpi-label">Outstanding</span><span class="ars-kpi-dot"></span></div>
					<div class="ars-kpi-value" id="ars-k-out">$0.00</div>
				</div>
				<div class="ars-kpi ars-kpi--rate">
					<div class="ars-kpi-top"><span class="ars-kpi-label">Collected</span><span class="ars-kpi-dot"></span></div>
					<div class="ars-kpi-value" id="ars-k-rate">0%</div>
				</div>
				<div class="ars-kpi ars-kpi--cust">
					<div class="ars-kpi-top"><span class="ars-kpi-label">Customers</span><span class="ars-kpi-dot"></span></div>
					<div class="ars-kpi-value" id="ars-k-cust">0</div>
				</div>
			</section>

			<div class="ars-toolbar">
				<div class="ars-field ars-search-field">
					<span class="ars-search-icon">&#128269;</span>
					<input type="text" id="ars-search" class="ars-input" placeholder="Search customer" />
				</div>
				<select id="ars-company" class="ars-select" title="Company">
					<option value="">All Companies</option>
				</select>
				<select id="ars-status" class="ars-select ars-summary-only" title="Payment status">
					<option value="all">All statuses</option>
					<option value="unpaid">Unpaid</option>
					<option value="partial">Partially paid</option>
					<option value="paid">Fully paid</option>
				</select>
				<div class="ars-daterange" title="Invoice date range">
					<input type="date" id="ars-from" />
					<span class="ars-arrow">&rarr;</span>
					<input type="date" id="ars-to" />
				</div>
				<select id="ars-sort" class="ars-select ars-summary-only" title="Sort by">
					<option value="outstanding">Outstanding (high &rarr; low)</option>
					<option value="invoice_total">Invoice total (high &rarr; low)</option>
					<option value="paid">Paid (high &rarr; low)</option>
					<option value="invoices">Most invoices</option>
					<option value="customer">Customer (A &rarr; Z)</option>
				</select>
				<div class="ars-toolbar-sep"></div>
				<button id="ars-clear" class="ars-ghost-btn">Clear filters</button>
			</div>

			<div id="ars-view-summary">
				<div class="ars-table-card" id="ars-table-card">
					<div class="ars-scroll">
						<table class="ars-table">
							<thead>
								<tr>
									<th>Customer</th>
									<th class="ars-c"># Inv</th>
									<th class="ars-r">Invoice Total</th>
									<th class="ars-r">Paid Amount</th>
									<th class="ars-r">Outstanding</th>
									<th class="ars-c">Status</th>
								</tr>
							</thead>
							<tbody id="ars-body"></tbody>
							<tfoot id="ars-foot"></tfoot>
						</table>
					</div>
					<div class="ars-empty" id="ars-empty" style="display:none;">No matching customers.</div>
				</div>
				<div class="ars-footer-row" id="ars-count"></div>
			</div>

			<div id="ars-view-ledgers" style="display:none;">
				<div id="ars-ledgers"></div>
				<div class="ars-empty" id="ars-ledgers-empty" style="display:none;">No matching customers.</div>
			</div>
		</div>
	`);

	function money(v) {
		return '$' + Number(v || 0).toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 });
	}
	function esc(s) { return frappe.utils.escape_html(s == null ? '' : String(s)); }

	function status_of(r) {
		var out = flt(r.outstanding), paid = flt(r.paid);
		if (out <= 0) return 'paid';
		if (paid > 0) return 'partial';
		return 'unpaid';
	}
	var STATUS_META = {
		paid:    { cls: 'ars-pill--paid',    label: 'Paid' },
		partial: { cls: 'ars-pill--partial', label: 'Partial' },
		unpaid:  { cls: 'ars-pill--unpaid',  label: 'Unpaid' }
	};

	// map a Frappe Sales Invoice status string to a pill style
	function invoice_pill(status) {
		var s = (status || '').toLowerCase();
		var cls = 'ars-pill--unpaid';
		if (s.indexOf('overdue') !== -1) cls = 'ars-pill--overdue';
		else if (s === 'paid') cls = 'ars-pill--paid';
		else if (s.indexOf('partly') !== -1) cls = 'ars-pill--partial';
		else if (s.indexOf('return') !== -1 || s.indexOf('credit') !== -1) cls = 'ars-pill--muted';
		return '<span class="ars-pill ' + cls + '">' + esc(status || '') + '</span>';
	}

	// render a Sales Order / Delivery Note reference cell
	function ref_cell(list, doctype) {
		list = list || [];
		if (!list.length) return '<td class="ars-c ars-muted">&mdash;</td>';
		var html = list.map(function (n) {
			return '<a class="ars-inv-link" href="/app/' + doctype + '/' + encodeURIComponent(n) + '">' + esc(n) + '</a>';
		}).join(', ');
		return '<td>' + html + '</td>';
	}

	// ---------- summary view ----------
	function visible_rows() {
		var q = (state.search || '').toLowerCase();
		var rows = state.rows.filter(function (r) {
			if (q && (r.customer || '').toLowerCase().indexOf(q) === -1) return false;
			if (state.status !== 'all' && status_of(r) !== state.status) return false;
			return true;
		});
		var s = state.sort;
		rows.sort(function (a, b) {
			if (s === 'customer') return (a.customer || '').localeCompare(b.customer || '');
			if (s === 'invoices') return (b.invoices || 0) - (a.invoices || 0);
			return flt(b[s]) - flt(a[s]);
		});
		return rows;
	}

	function render_summary() {
		var rows = visible_rows();
		var $body = $('#ars-body').empty();

		rows.forEach(function (r) {
			var out = flt(r.outstanding), paid = flt(r.paid), total = flt(r.invoice_total);
			var pct = total > 0 ? Math.min(100, Math.round((paid / total) * 100)) : 0;
			var st = STATUS_META[status_of(r)];
			var link = '/app/customer/' + encodeURIComponent(r.customer);
			$body.append(
				'<tr>' +
				'<td class="ars-cust-cell">' +
					'<a class="ars-cust-name" href="' + link + '">' + esc(r.customer) + '</a>' +
					'<div class="ars-progress"><div class="ars-progress-fill" style="width:' + pct + '%"></div></div>' +
				'</td>' +
				'<td class="ars-c ars-muted">' + (r.invoices || 0) + '</td>' +
				'<td class="ars-r">' + money(total) + '</td>' +
				'<td class="ars-r ars-paid">' + money(paid) + '</td>' +
				'<td class="ars-r ars-out ' + (out > 0 ? '' : 'ars-zero') + '">' + money(out) + '</td>' +
				'<td class="ars-c"><span class="ars-pill ' + st.cls + '">' + st.label + '</span></td>' +
				'</tr>'
			);
		});
		$('#ars-empty').toggle(rows.length === 0);

		var ft = rows.reduce(function (a, r) {
			a.invoices += (r.invoices || 0); a.invoice_total += flt(r.invoice_total);
			a.paid += flt(r.paid); a.outstanding += flt(r.outstanding); return a;
		}, { invoices: 0, invoice_total: 0, paid: 0, outstanding: 0 });

		$('#ars-foot').html(rows.length ?
			('<tr class="ars-total"><td>TOTAL</td>' +
			 '<td class="ars-c">' + ft.invoices + '</td>' +
			 '<td class="ars-r">' + money(ft.invoice_total) + '</td>' +
			 '<td class="ars-r">' + money(ft.paid) + '</td>' +
			 '<td class="ars-r">' + money(ft.outstanding) + '</td><td></td></tr>') : '');

		update_kpis(ft.invoice_total, ft.paid, ft.outstanding, rows.length);
		$('#ars-count').text('Showing ' + rows.length + ' of ' + state.rows.length + ' customers');
	}

	// ---------- ledgers view ----------
	function render_ledgers() {
		var groups = (state.ledgers && state.ledgers.groups) || [];
		var q = (state.search || '').toLowerCase();
		if (q) groups = groups.filter(function (g) { return (g.customer || '').toLowerCase().indexOf(q) !== -1; });

		var $host = $('#ars-ledgers').empty();
		$('#ars-ledgers-empty').toggle(groups.length === 0);

		var tot = { invoice_total: 0, paid: 0, outstanding: 0 };

		groups.forEach(function (g) {
			var t = g.totals || {};
			tot.invoice_total += flt(t.invoice_total); tot.paid += flt(t.paid); tot.outstanding += flt(t.outstanding);

			var body = '';
			(g.invoices || []).forEach(function (inv) {
				var link = '/app/sales-invoice/' + encodeURIComponent(inv.name);
				body +=
					'<tr>' +
					'<td><a class="ars-inv-link" href="' + link + '">' + esc(inv.name) + '</a></td>' +
					ref_cell(inv.sales_order, 'sales-order') +
					'<td class="ars-muted">' + esc(inv.company) + '</td>' +
					'<td class="ars-c ars-muted">' + esc(inv.posting_date) + '</td>' +
					'<td class="ars-c ars-muted">' + esc(inv.due_date) + '</td>' +
					'<td class="ars-r">' + money(inv.invoice_total) + '</td>' +
					'<td class="ars-r ars-paid">' + money(inv.paid) + '</td>' +
					'<td class="ars-r ars-out ' + (flt(inv.outstanding) > 0 ? '' : 'ars-zero') + '">' + money(inv.outstanding) + '</td>' +
					'<td class="ars-c">' + invoice_pill(inv.status) + '</td>' +
					'</tr>';
			});

			$host.append(
				'<div class="ars-ledger">' +
					'<div class="ars-ledger-head">' +
						'<div class="ars-ledger-title">' + esc(g.customer) +
							'<span class="ars-ledger-count">' + g.count + ' invoice' + (g.count === 1 ? '' : 's') + '</span></div>' +
						'<div class="ars-ledger-out">Outstanding <b>' + money(t.outstanding) + '</b></div>' +
					'</div>' +
					'<div class="ars-scroll"><table class="ars-ltable">' +
						'<thead><tr>' +
							'<th>Invoice #</th><th>Sales Order</th><th>Company</th>' +
							'<th class="ars-c">Posting Date</th><th class="ars-c">Due Date</th>' +
							'<th class="ars-r">Invoice Total</th><th class="ars-r">Paid Amount</th>' +
							'<th class="ars-r">Outstanding</th><th class="ars-c">Status</th>' +
						'</tr></thead>' +
						'<tbody>' + body + '</tbody>' +
						'<tfoot><tr class="ars-ltotal">' +
							'<td colspan="5">TOTAL</td>' +
							'<td class="ars-r">' + money(t.invoice_total) + '</td>' +
							'<td class="ars-r">' + money(t.paid) + '</td>' +
							'<td class="ars-r ars-out">' + money(t.outstanding) + '</td><td></td>' +
						'</tr></tfoot>' +
					'</table></div>' +
				'</div>'
			);
		});

		update_kpis(tot.invoice_total, tot.paid, tot.outstanding, groups.length);
	}

	function update_kpis(invoice_total, paid, outstanding, customers) {
		var rate = invoice_total > 0 ? Math.round((paid / invoice_total) * 100) : 0;
		$('#ars-k-inv').text(money(invoice_total));
		$('#ars-k-paid').text(money(paid));
		$('#ars-k-out').text(money(outstanding));
		$('#ars-k-rate').text(rate + '%');
		$('#ars-k-cust').text(customers);
	}

	function render_active() {
		if (state.view === 'ledgers') render_ledgers();
		else render_summary();
	}

	// ---------- data loading ----------
	function set_loading(on) {
		$('#ars-table-card').toggleClass('ars-loading', on);
		$('#ars-ledgers').toggleClass('ars-loading', on);
		$('#ars-refresh-ic').toggleClass('ars-spin', on);
	}

	function load_summary(then) {
		set_loading(true);
		frappe.call({
			method: METHOD + '.get_ar_summary',
			args: { company: state.company, from_date: state.from_date, to_date: state.to_date },
			callback: function (r) {
				if (!r.message) return;
				state.rows = r.message.rows || [];
				state.prepared = r.message.prepared || '';
				$('#ars-prepared').text('Prepared ' + state.prepared + ' \u2014 Accounts Receivable');
				if (then) then();
			},
			always: function () { set_loading(false); }
		});
	}

	function load_ledgers(then) {
		set_loading(true);
		frappe.call({
			method: METHOD + '.get_ledgers',
			args: { company: state.company, from_date: state.from_date, to_date: state.to_date },
			callback: function (r) {
				if (r.message) state.ledgers = r.message;
				if (then) then();
			},
			always: function () { set_loading(false); }
		});
	}

	function reload() {
		if (state.view === 'ledgers') load_ledgers(render_ledgers);
		else load_summary(render_summary);
	}

	function switch_view(view) {
		state.view = view;
		$('#ars-seg .ars-seg-btn').removeClass('is-active')
			.filter('[data-view="' + view + '"]').addClass('is-active');
		$('#ars-view-summary').toggle(view === 'summary');
		$('#ars-view-ledgers').toggle(view === 'ledgers');
		$('.ars-summary-only').toggle(view === 'summary');

		if (view === 'ledgers') {
			if (state.ledgers) render_ledgers(); else load_ledgers(render_ledgers);
		} else {
			if (state.rows.length) render_summary(); else load_summary(render_summary);
		}
	}

	// ---------- events ----------
	$('#ars-seg').on('click', '.ars-seg-btn', function () { switch_view($(this).data('view')); });

	function server_filter_changed() { state.ledgers = null; state.rows = []; reload(); }
	$('#ars-company').on('change', function () { state.company = $(this).val(); server_filter_changed(); });
	$('#ars-from').on('change', function () { state.from_date = $(this).val(); server_filter_changed(); });
	$('#ars-to').on('change', function () { state.to_date = $(this).val(); server_filter_changed(); });
	$('#ars-refresh').on('click', server_filter_changed);

	$('#ars-search').on('input', function () { state.search = $(this).val(); render_active(); });
	$('#ars-status').on('change', function () { state.status = $(this).val(); render_summary(); });
	$('#ars-sort').on('change', function () { state.sort = $(this).val(); render_summary(); });

	$('#ars-clear').on('click', function () {
		state.search = ''; state.status = 'all'; state.sort = 'outstanding';
		state.company = ''; state.from_date = ''; state.to_date = ''; state.ledgers = null; state.rows = [];
		$('#ars-search').val(''); $('#ars-status').val('all'); $('#ars-sort').val('outstanding');
		$('#ars-company').val(''); $('#ars-from').val(''); $('#ars-to').val('');
		reload();
	});

	$('#ars-export').on('click', function () {
		open_url_post('/api/method/' + METHOD + '.export_ar_summary', {
			company: state.company || '', from_date: state.from_date || '', to_date: state.to_date || ''
		});
	});

	load_summary(render_summary);
};

File: cannabis_management/cannabis_management/page/ar_summary/ar_summary.py
"""Accounts Receivable — summary + per-customer ledgers.

Intercompany / internal-transfer invoices (is_internal_customer = 1, which Frappe
shows with the "Internal Transfer" status) are excluded from every view and export.

The ledger carries the Sales Order and Delivery Note the invoice was created against.
Outstanding balances are shown in red (page and export). The Excel export carries two
sheets: "Accounts Receivable" (summary) and "Customer Ledgers" (boxed blocks).
"""

import io

import frappe
from frappe.utils import flt, formatdate, get_url, nowdate

# ---- summary sheet colours -------------------------------------------------
YELLOW = "FFE699"   # Invoice Total
GREEN = "C6EFCE"    # Paid / settled outstanding
RED = "FFC7CE"      # outstanding still owed
HEADER = "DDEBF7"   # header row
RED_FONT = "C00000" # red text for outstanding balances

# ---- ledger sheet colours (match the Customer Ledgers screenshot) ----------
NAVY = "1F3864"     # per-customer title band
COLHDR = "D9D9D9"   # column-header row
AMT = "E2EFDA"      # amount cells in data rows (light green)
TOTAL_YELLOW = "FFE699"


def _conditions(company=None, from_date=None, to_date=None):
    """Shared WHERE clause. Always drops internal / intercompany transfers."""
    conditions = ["docstatus = 1", "IFNULL(is_internal_customer, 0) = 0"]
    values = {}
    if company:
        conditions.append("company = %(company)s")
        values["company"] = company
    if from_date:
        conditions.append("posting_date >= %(from_date)s")
        values["from_date"] = from_date
    if to_date:
        conditions.append("posting_date <= %(to_date)s")
        values["to_date"] = to_date
    return " AND ".join(conditions), values


def _summary_rows(company=None, from_date=None, to_date=None):
    where, values = _conditions(company, from_date, to_date)
    return frappe.db.sql(
        f"""
        SELECT customer,
               COUNT(*) AS invoices,
               SUM(grand_total) AS invoice_total,
               SUM(grand_total - outstanding_amount) AS paid,
               SUM(outstanding_amount) AS outstanding
        FROM `tabSales Invoice`
        WHERE {where}
        GROUP BY customer
        HAVING invoice_total > 0
        ORDER BY outstanding DESC, invoice_total DESC
        """,
        values,
        as_dict=True,
    )


def _attach_references(invoices):
    """Attach each invoice's Sales Order(s) — the SO it was created against, read
    from the invoice item rows — and the Delivery Note(s) raised against those
    Sales Orders (NOT the delivery_note stamped on the invoice item)."""
    names = [i["name"] for i in invoices]
    if not names:
        return

    # 1. Sales Order(s) each invoice was created against.
    rows = frappe.db.sql(
        """
        SELECT parent, sales_order
        FROM `tabSales Invoice Item`
        WHERE parent IN %(names)s AND IFNULL(sales_order, '') != ''
        """,
        {"names": tuple(names)},
        as_dict=True,
    )
    so_map, all_sos = {}, set()
    for r in rows:
        so_map.setdefault(r.parent, set()).add(r.sales_order)
        all_sos.add(r.sales_order)

    # 2. Delivery Note(s) raised against those Sales Orders.
    so_to_dn = {}
    if all_sos:
        dn_rows = frappe.db.sql(
            """
            SELECT against_sales_order AS so, parent AS dn
            FROM `tabDelivery Note Item`
            WHERE against_sales_order IN %(sos)s AND docstatus = 1
            """,
            {"sos": tuple(all_sos)},
            as_dict=True,
        )
        for r in dn_rows:
            so_to_dn.setdefault(r.so, set()).add(r.dn)

    # 3. Payment Entry(ies) made against each invoice, or against its Sales Order
    #    (advances). Read from Payment Entry Reference; submitted payments only.
    inv_pe, so_pe = {}, {}
    pe_rows = frappe.db.sql(
        """
        SELECT per.reference_doctype AS rdt, per.reference_name AS rname, per.parent AS pe
        FROM `tabPayment Entry Reference` per
        INNER JOIN `tabPayment Entry` pe ON pe.name = per.parent AND pe.docstatus = 1
        WHERE (per.reference_doctype = 'Sales Invoice' AND per.reference_name IN %(inv)s)
           OR (per.reference_doctype = 'Sales Order' AND per.reference_name IN %(sos)s)
        """,
        {"inv": tuple(names), "sos": tuple(all_sos) if all_sos else ("",)},
        as_dict=True,
    )
    for r in pe_rows:
        (inv_pe if r.rdt == "Sales Invoice" else so_pe).setdefault(r.rname, set()).add(r.pe)

    for inv in invoices:
        sos = so_map.get(inv["name"], set())
        inv["sales_order"] = sorted(sos)
        dns = set()
        for so in sos:
            dns |= so_to_dn.get(so, set())
        pes = set(inv_pe.get(inv["name"], set()))
        for so in sos:
            pes |= so_pe.get(so, set())
        inv["payment_entries"] = sorted(pes)
        inv["delivery_note"] = sorted(dns)


def _ledger_groups(company=None, from_date=None, to_date=None):
    """Per-invoice detail grouped by customer, groups sorted by outstanding
    balance (highest first), matching the reference sheet."""
    where, values = _conditions(company, from_date, to_date)
    invoices = frappe.db.sql(
        f"""
        SELECT name, customer, company, posting_date, due_date,
               grand_total AS invoice_total,
               (grand_total - outstanding_amount) AS paid,
               outstanding_amount AS outstanding,
               status
        FROM `tabSales Invoice`
        WHERE {where}
        ORDER BY customer, posting_date, name
        """,
        values,
        as_dict=True,
    )
    _attach_references(invoices)

    groups = {}
    for inv in invoices:
        groups.setdefault(inv.customer, []).append(inv)

    result = []
    for customer, invs in groups.items():
        totals = {
            "invoice_total": sum(flt(i.invoice_total) for i in invs),
            "paid": sum(flt(i.paid) for i in invs),
            "outstanding": sum(flt(i.outstanding) for i in invs),
        }
        if totals["invoice_total"] <= 0:  # keep in step with the summary
            continue
        result.append({
            "customer": customer,
            "invoices": invs,
            "count": len(invs),
            "totals": totals,
        })

    result.sort(key=lambda g: g["totals"]["outstanding"], reverse=True)
    return result


@frappe.whitelist()
def get_ar_summary(company=None, from_date=None, to_date=None):
    rows = _summary_rows(company, from_date, to_date)
    totals = {
        "invoices": sum(r.invoices for r in rows),
        "invoice_total": sum(flt(r.invoice_total) for r in rows),
        "paid": sum(flt(r.paid) for r in rows),
        "outstanding": sum(flt(r.outstanding) for r in rows),
    }
    return {
        "rows": rows,
        "totals": totals,
        "prepared": formatdate(nowdate(), "MMMM dd, yyyy"),
    }


@frappe.whitelist()
def get_ledgers(company=None, from_date=None, to_date=None):
    groups = _ledger_groups(company, from_date, to_date)
    for g in groups:
        for inv in g["invoices"]:
            inv["posting_date"] = formatdate(inv.posting_date, "yyyy-MM-dd") if inv.posting_date else ""
            inv["due_date"] = formatdate(inv.due_date, "yyyy-MM-dd") if inv.due_date else ""
    return {
        "groups": groups,
        "prepared": formatdate(nowdate(), "MMMM dd, yyyy"),
    }


# ===========================================================================
# Excel export — two sheets
# ===========================================================================

def _build_summary_sheet(ws, rows, prepared):
    from openpyxl.styles import Alignment, Border, Font, PatternFill, Side

    thin = Side(style="thin", color="BFBFBF")
    border = Border(left=thin, right=thin, top=thin, bottom=thin)
    money = "$#,##0.00"
    center = Alignment(horizontal="center")
    right = Alignment(horizontal="right")

    def fill(hexcode):
        return PatternFill("solid", fgColor=hexcode)

    ws.title = "Accounts Receivable"
    ws.merge_cells("A1:E1")
    ws["A1"] = f"Prepared {prepared} — Accounts Receivable"
    ws["A1"].font = Font(italic=True, size=10, color="595959")

    headers = ["Customer", "# Invoices", "Invoice Total", "Paid Amount", "Outstanding"]
    hr = 3
    for c, h in enumerate(headers, start=1):
        cell = ws.cell(row=hr, column=c, value=h)
        cell.font = Font(bold=True)
        cell.fill = fill(HEADER)
        cell.border = border
        cell.alignment = center if c > 1 else Alignment(horizontal="left")

    r = hr + 1
    for row in rows:
        outstanding = flt(row.outstanding)
        paid = flt(row.paid)
        cust = ws.cell(row=r, column=1, value=row.customer)
        inv = ws.cell(row=r, column=2, value=row.invoices)
        it = ws.cell(row=r, column=3, value=flt(row.invoice_total))
        pd = ws.cell(row=r, column=4, value=paid)
        out = ws.cell(row=r, column=5, value=outstanding)

        for cell in (cust, inv, it, pd, out):
            cell.border = border
        inv.alignment = center
        for cell in (it, pd, out):
            cell.number_format = money
            cell.alignment = right

        it.fill = fill(YELLOW)
        if paid > 0:
            pd.fill = fill(GREEN)
        # outstanding owed -> red; settled -> green
        if outstanding > 0:
            out.fill = fill(RED)
            out.font = Font(color=RED_FONT, bold=True)
        else:
            out.fill = fill(GREEN)
        r += 1

    tr = r
    ws.cell(row=tr, column=1, value="TOTAL").font = Font(bold=True)
    ws.cell(row=tr, column=2, value=sum(x.invoices for x in rows)).alignment = center
    for col, key in ((3, "invoice_total"), (4, "paid"), (5, "outstanding")):
        cell = ws.cell(row=tr, column=col, value=sum(flt(x[key]) for x in rows))
        cell.number_format = money
        cell.font = Font(bold=True, color=RED_FONT) if col == 5 else Font(bold=True)
        cell.alignment = right
    for c in range(1, 6):
        ws.cell(row=tr, column=c).border = border

    for col, w in {"A": 34, "B": 11, "C": 16, "D": 16, "E": 16}.items():
        ws.column_dimensions[col].width = w
    ws.freeze_panes = "A4"


def _ref_cell(ws, row, col, names, doctype, border, fill_amt, left):
    """Write a Sales Order / Delivery Note reference cell. Hyperlinks a single
    reference; comma-joins multiples as plain text."""
    from openpyxl.styles import Font

    names = names or []
    cell = ws.cell(row=row, column=col, value=", ".join(names))
    cell.border = border
    cell.alignment = left
    cell.fill = fill_amt
    if len(names) == 1:
        cell.hyperlink = f"{get_url()}/app/{doctype}/{names[0]}"
        cell.font = Font(color="0563C1", underline="single")
    return cell


def _build_ledger_sheet(ws, groups, prepared):
    from openpyxl.styles import Alignment, Border, Font, PatternFill, Side

    thin = Side(style="thin", color="BFBFBF")
    border = Border(left=thin, right=thin, top=thin, bottom=thin)
    money = "$#,##0.00"
    center = Alignment(horizontal="center")
    right = Alignment(horizontal="right")
    left = Alignment(horizontal="left")

    def fill(hexcode):
        return PatternFill("solid", fgColor=hexcode)

    fill_amt = fill(AMT)
    fill_ref = fill(AMT)  # light tint behind reference cells too (kept subtle/white-ish)

    ws.title = "Customer Ledgers"
    LAST = 10  # A:J
    ws.merge_cells(start_row=1, start_column=1, end_row=1, end_column=LAST)
    ws["A1"] = (f"Prepared {prepared} — each customer is its own boxed block, "
                "sorted by outstanding balance (highest first)")
    ws["A1"].font = Font(italic=True, size=10, color="595959")

    headers = ["Invoice #", "Sales Order", "Payment",
               "Company", "Posting Date", "Due Date", "Invoice Total",
               "Paid Amount", "Outstanding", "Status"]

    r = 3
    for g in groups:
        n = g["count"]
        label = f"{g['customer']}   ({n} invoice{'s' if n != 1 else ''})"

        # customer title band (navy, white, merged A:J)
        ws.merge_cells(start_row=r, start_column=1, end_row=r, end_column=LAST)
        band = ws.cell(row=r, column=1, value=label)
        band.font = Font(bold=True, color="FFFFFF", size=11)
        band.fill = fill(NAVY)
        band.alignment = left
        for c in range(1, LAST + 1):
            ws.cell(row=r, column=c).border = border
        r += 1

        # column headers
        for c, h in enumerate(headers, start=1):
            cell = ws.cell(row=r, column=c, value=h)
            cell.font = Font(bold=True)
            cell.fill = fill(COLHDR)
            cell.border = border
            cell.alignment = center if c >= 5 else left
        r += 1

        # invoice rows
        for inv in g["invoices"]:
            name_cell = ws.cell(row=r, column=1, value=inv["name"])
            name_cell.hyperlink = f"{get_url()}/app/sales-invoice/{inv['name']}"
            name_cell.font = Font(color="0563C1", underline="single")
            name_cell.alignment = left
            name_cell.border = border

            _ref_cell(ws, r, 2, inv.get("sales_order"), "sales-order", border, PatternFill(), left)
            _ref_cell(ws, r, 3, inv.get("payment_entries"), "payment-entry", border, PatternFill(), left)

            ws.cell(row=r, column=4, value=inv["company"]).alignment = left
            ws.cell(row=r, column=5,
                    value=formatdate(inv["posting_date"], "yyyy-MM-dd") if inv["posting_date"] else "").alignment = center
            ws.cell(row=r, column=6,
                    value=formatdate(inv["due_date"], "yyyy-MM-dd") if inv["due_date"] else "").alignment = center

            it = ws.cell(row=r, column=7, value=flt(inv["invoice_total"]))
            pd = ws.cell(row=r, column=8, value=flt(inv["paid"]))
            out = ws.cell(row=r, column=9, value=flt(inv["outstanding"]))
            for cell in (it, pd):
                cell.number_format = money
                cell.alignment = right
                cell.fill = fill_amt
            out.number_format = money
            out.alignment = right
            # outstanding owed -> red; settled -> green
            if flt(inv["outstanding"]) > 0:
                out.fill = fill(RED)
                out.font = Font(color=RED_FONT, bold=True)
            else:
                out.fill = fill(GREEN)

            ws.cell(row=r, column=10, value=inv["status"]).alignment = center

            for c in range(1, LAST + 1):
                ws.cell(row=r, column=c).border = border
            r += 1

        # per-customer TOTAL row (yellow); outstanding total in red
        t = g["totals"]
        ws.cell(row=r, column=1, value="TOTAL")
        for c in range(1, 7):
            ws.cell(row=r, column=c).fill = fill(TOTAL_YELLOW)
            ws.cell(row=r, column=c).font = Font(bold=True)
        for col, key in ((7, "invoice_total"), (8, "paid"), (9, "outstanding")):
            cell = ws.cell(row=r, column=col, value=flt(t[key]))
            cell.number_format = money
            cell.alignment = right
            cell.fill = fill(TOTAL_YELLOW)
            cell.font = Font(bold=True, color=RED_FONT) if col == 9 else Font(bold=True)
        ws.cell(row=r, column=10).fill = fill(TOTAL_YELLOW)
        for c in range(1, LAST + 1):
            ws.cell(row=r, column=c).border = border
        r += 2  # blank spacer between customers

    widths = {"A": 22, "B": 20, "C": 20, "D": 24, "E": 13,
              "F": 13, "G": 14, "H": 14, "I": 14, "J": 13}
    for col, w in widths.items():
        ws.column_dimensions[col].width = w
    ws.freeze_panes = "A3"


@frappe.whitelist()
def export_ar_summary(company=None, from_date=None, to_date=None):
    from openpyxl import Workbook

    rows = _summary_rows(company, from_date, to_date)
    groups = _ledger_groups(company, from_date, to_date)
    prepared = formatdate(nowdate(), "MMMM dd, yyyy")

    wb = Workbook()
    _build_summary_sheet(wb.active, rows, prepared)
    _build_ledger_sheet(wb.create_sheet(), groups, prepared)

    bio = io.BytesIO()
    wb.save(bio)
    bio.seek(0)
    frappe.response["filename"] = f"Accounts Receivable {prepared}.xlsx"
    frappe.response["filecontent"] = bio.read()
    frappe.response["type"] = "binary"